rfcs/0041-new-transaction-model.md

```diff
@@ -0,0 +1,183 @@
+
+Currently, the transitions in our state machine are individual transactions. This proposes extending that with the transitions
+
+```
+type transition =
+  | Transaction of Transaction.t
+  | Step_or_start_party_sequence of step_or_start
+
+type step_or_start =
+  | Step
+  | Start of party list
+```
+
+It remains to explain how to execute a "party" as a state transition.
+In pseudocaml/snarky, it will work as follows
+
+```ocaml
+let apply_step_or_start
+  (e : execution_state) (instr : step_or_start)
+  : execution_state
+  =
+  let local_state =
+    match e.current_transaction_local_state, instr with
+    | None, Step
+    | Some _, Start _ -> assert false
+    | None, Start ps ->
+      {default_local_state with parties=ps; id= hash ps}
+    | Some s, Step -> s
+  in
+  let {authorization; predicate; update}, remaining =
+  	Non_empty_list.uncons local_state.remaining_parties
+  in
+  let a, merkle_path = exists ~request:Party_account in
+  assert (implied_root a merkle_path e.ledger_hash = e.ledger_hash) ;
+  assert (verify_authorization authorization a s.id) ;
+  assert (verify_predicate predicate a) ;
+  assert (auth_sufficient_given_permissions a.permissions authorization update) ;
+  let fee_excess_change =
+    match s.current_token_id with
+    | None -> a.token_id
+    | Some curr ->
+      if curr == a.token_id
+      then Currency.Fee.zero
+      else (
+        (* If we are switching tokens, then we cannot have printed money out of thin air. *)
+        assert (s.excess >= 0);
+        if curr == Token_id.default
+        then s.excess
+        else 0 (* burn the excess of this non-default token *)
+      )
+  in    
+  let a' = perform_update update a in
+  let excess = current_transaction_local_state.excess + (a.balance - a'.balance) in
+  let new_ledger_hash = implied_root a' merkle_path in
+  match remaining with
+  | [] ->
+  	assert (excess >= 0);
+  	let fee_excess_change =
+  	  fee_excess_change +
+  	  if a.token_id == Token_id.default then excess else 0
+  	in
+  	{ e with current_transaction_local_state= None
+  	; ledger_hash= new_ledger_hash
+  	; fee_excess= e.fee_excess + fee_excess_change }
+  | _::_ ->
+  	{ e with current_transaction_local_state=
+  	  Some
+  	    { local_state with parties= remaining
+  	    ; excess= local_state.excess + excess }
+  	; ledger_hash= new_ledger_hash 
+  	; fee_excess= e.fee_excess + excess_change }
+```
+
+### How this would boil down into "base" transaction SNARKs
+The idea would be to have 3 new base transaction SNARKs corresponding to the 3 forms of authentication. Each would implement the above `apply_step_or_start` function but with the `verify_authorization` specialized to either signature verification, SNARK verification, or nothing.
+
+
+Under this model, executing a transaction works as follows. Let `t = [t1; t2; t3]`
+
+### Fees and proofs required
+
+Instead of 2 proofs per transaction as is required now, we will switch to 2 proofs per "party".
```

Comment:
This is a good point. We will need to think carefully about our block production logic in this transaction model.

Reply:
> If I understood correctly, all the parties within a transaction need to be included in one block. Like, including a payment in a block would mean including both the (or 3) parties in the scan state. Are we going to cap the number parties based on the scan state size?

Just making sure I understand this correctly: do you mean all the parties need to be included in the same block or in the same tree? All the parties within a transaction would be in the same block, since a block contains a list of transactions and each transaction is a list of parties. I think we can actually avoid the requirement of all parties being in the same tree by having the ledger for the currently executing transaction as a kind of "hypothetical" state that gets committed once all the parties in the transaction have finished.

> Furthermore, seems like all the parties of a transaction and the ones that settles the fees need to be in the same tree so that the net fee excess is zero.

Do you mean the corresponding fee transfers should be in the same tree? Actually how does this work now (isn't it the same problem)?

Reply:
> If I understood correctly, all the parties within a transaction need to be included in one block. Like, including a payment in a block would mean including both the (or 3) parties in the scan state. Are we going to cap the number parties based on the scan state size?

Good question. Just making sure I understand this correctly: do you mean all the parties need to be included in the same block or in the same tree? All the parties within a transaction would be in the same block, since a block contains a list of transactions and each transaction is a list of parties. I think we can actually avoid the requirement of all parties being in the same tree by having the ledger for the currently executing transaction as a kind of "hypothetical" state that gets committed once all the parties in the transaction have finished.

> Furthermore, seems like all the parties of a transaction and the ones that settles the fees need to be in the same tree so that the net fee excess is zero.

Do you mean the corresponding fee transfers should be in the same tree? Actually how does this work now (isn't it the same problem)?

Reply:
>  do you mean all the parties need to be included in the same block or in the same tree? 
The same block. This was to confirm if we'll need to cap the number of parties per transaction based on the scan state size.

>Do you mean the corresponding fee transfers should be in the same tree? Actually how does this work now (isn't it the same problem)?

The parties of a transaction and the corresponding fee transfers should be in the same tree for the fee excess to be zero in the final merge proof. This is the case even now, just that it is always one transaction per leaf. So to add a transaction, in the worst case we'll need 3 leaves (1 transaction, 2 fee transfers). If the number of leaves in a tree <= 2 and we can't add any transaction because we can't fit in the required fee transfers, then we fill them with coinbase transactions (split the coinbase amount among two transactions). Now with each party occupying a leaf and there being arbitrary number of parties, it is not very clear what the optimal solution would be